lib/auth/auth.go

```diff
@@ -655,7 +655,7 @@ func (a *Server) GenerateHostCert(hostPublicKey []byte, hostID, nodeName string,
 	}
 
 	// get the certificate authority that will be signing the public key of the host
-	ca, err := a.GetCertAuthority(context.TODO(), types.CertAuthID{
+	ca, err := a.Services.GetCertAuthority(context.TODO(), types.CertAuthID{
```

Comment:
This is a little confusing - how do we know when we have to delegate to `Services` and when we can call `a`'s method directly?

Reply:
imo, we should look to move away from embedding `Cache` and `Services` into the `auth.Server`. It would probably be clearer what's going on in most places if we had to use the fully qualified names.

Reply:
@zmb3 any flow that involves reading a resource, changing it and upserting it; any flow that updates or deletes something based on the result of a read; any automatic flow that's triggered by a change and expects to read or use said change immediately

This complication is exacerbated by the fact that some non-singleton resources will read through the cache if a resource is missing (but not if it has been changed, or when fetching multiple resources at once).


@strideynet 
That's a discussion that @codingllama and I already started in the original PR.

Removing the embedding would be cool but requires either a lot of explicit method delegation or about 950 changes in call sites all over the code; if we want to go that route, we need automation.

Our interfaces are also not great for that, as there's nothing that lets you split read methods that could use the cache, read methods that can't, and write methods.